### DISCUSSION TITLE
M1 mac not able to run

### DISCUSSION THREAD

**Discussion by murugaratham** [datalust/seq-tickets#1189] [General]
```
Function not implemented (See inner exception for details.) ---> System.IO.IOException: Function not implemented
   at System.IO.FileSystemWatcher.StartRaisingEvents()
```

Similar issue filed [here](https://github.com/docker/for-mac/issues/5328) for docker/for-mac

I think it would be a quick fix if datalust can build Arm64v8 images, since the dockerfile now seems to be close sourced

**Comment by nblumhardt** [comment]
Thanks for the note, @murugaratham. Seq relies on a lot of native code, and so there's quite a bit more for us to do than just rebuilding the Docker image, unfortunately.

We're expecting to begin M1/ARM support in earnest once .NET 6 ships (November 2021).

**Reply by murugaratham** [reply]
Thanks for your quick response, and sorry for under estimating the scope of work. Will wait for .net 6!

**Reply by nblumhardt** [reply]
No worries! We're looking forward to getting stuck into this :-)

**Comment by murugaratham** [comment]
@nblumhardt are there any preview/prerelease channel where we can use for dev purpose? We can also provide issues/bugs that we face along the way

**Reply by KodrAus** [reply]
Hi @murugaratham :wave:

It looks like a fresh preview of the .NET 6 SDK with support for Apple Silicon just landed last week. I've successfully had Seq building and running on my local M1 machine (and a Raspberry Pi 4 too for good measure) here with earlier daily builds of the .NET 6 SDK but there's still lots for us to do to get our CI and release processes able to support ARM builds before we can start releasing them.

I think around the time .NET 6 reaches a production-supported RTM release we'll be able to seriously consider shipping ARM builds of Seq. We don't _technically_ need to wait for .NET 6 to target ARM Linux, but it's easier for us to try handle all the ARM work together.

**Reply by murugaratham** [reply]
Alright, I’ll wait patiently then 🥲

**Comment by KodrAus** [comment]
You can now pull `datalust/seq:2021.3.6487-pre` on ARM-based environments and get yourself a working* non-mission-critical Seq instance :tada:

*: There are some very important caveats in this build for ARM (none of these apply to x86):

- We haven't got all the CI infrastructure in place to properly build and test ARM images, so the confidence that things work comes from me building and running Seq on my M1 Mac over the year rather than from heavy burn-in testing. That also means future ARM preview releases will depend on me putting a build together rather than automatically through our CI for now.
- Apps that use native binaries (like JSON Archive) won't run because we don't have ARM builds for them yet.
- The document metadata store is a new implementation that just stores line-delimited JSON updates on-disk. It doesn't do any file rolling or recovery yet so it's definitely not suitable for any serious workloads. That's not the event store where logs go though. It's only things like signals, alerts, charts etc that are stored in the document metadata store. The event store in this image is the same as on x86 and uses Neon intrinsics in place of SSE/AVX so is pretty battle-tested at this point.
- The .NET side of things is built using a recent daily build of the .NET 6 RC SDK. There may be bugs lurking either in the runtime, SDK libraries, or our usage of them that haven't surfaced yet.
- The Rust side of things is built using a recent nightly toolchain. There may be bugs lurking in the compiler or standard library that haven't surfaced yet.

Having said all that, I'm really excited to push this out! If anybody wants to give it a try I'd love to hear about how it goes. I'll post any future updates here until we can start shipping production-supported ARM builds for real.

**Reply by KodrAus** [reply]
Just pushed `datalust/seq:2021.3.6487-pre`

**Reply by KodrAus** [reply]
`datalust/seq:2021.3.6651` also works on ARM

**Reply by nathan-alden-sr** [reply]
Thanks so much for this. I, too, and trying to run Seq on my Raspberry Pi 4 so this is very useful.

**Reply by ricardoboss** [reply]
@KodrAus what about ARM/v7?

**Reply by KodrAus** [reply]
Hi @ricardoboss! To be completely honest I think armv7 probably isn't on the horizon for us, for a couple of reasons that really boil down to it being a major undertaking:

1. We've only ever built Seq for 64bit architectures to date, so porting to a 32bit one would require a lot of very careful consideration. We try not to make assumptions about the underlying platform but I'd expect there to be some very subtle issues unearthed to tease out.
2. We have a large native codebase that makes use of hardware intrinsics so porting to new architectures requires porting algorithms that use those intrinsics too.
3. Each new architecture we support comes with a burden in CI so that we can build, test, and ship binaries for it.

Since 32bit architectures aren't so common anymore, and Seq is a memory-intensive system, it's probably not something we'd be able to make enough of a priority.
